src/SPH.cpp

```diff
@@ -306,57 +316,63 @@ void SPH::update_positions() const
 	}
 }
 
-void SPH::correct_divergence_error(float* k_v_i, float* scalar_values, float* alpha)
+/*
+ * ViscousDFSPH, Algorithm 2
+ */
+void SPH::correct_divergence_error(float* dens_derive, float* scalar_values, float* alpha)
 {
-	float dens_i;
-	float div_i;
-	float sum_x, sum_y, sum_z;
-	float dx, dy, dz;
-	float kernel_gradient_x, kernel_gradient_y, kernel_gradient_z;
 	int neighbor_ind;
+	int iter = 0;
 
-	float dens_avg = calculate_stiffness(alpha, &m_particles.pred_vel, &m_particles.pos, m_delta_t, k_v_i, m_neighbor_data, scalar_values, m_particles.mass);
-
-	do 
+	float dens_derive_avg;
+	float k_v_i, k_v_j, div_i, div_j;
+	float pressure_acc_x, pressure_acc_y, pressure_acc_z;
+	float x, y, z;
+	float kernel_gradient_x, kernel_gradient_y, kernel_gradient_z;
+	
+	do
 	{
+		dens_derive_avg = calculate_derived_density(dens_derive, &m_particles.pred_vel, m_particles.mass, scalar_values, m_neighbor_data, &m_particles.pos);
+		
 		for (auto particle_ind = 0; particle_ind < D_NR_OF_PARTICLES; ++particle_ind)
 		{
-			dens_i = m_particles.dens[particle_ind];
 
-			assert(dens_i != 0.0f, "dens");
+			k_v_i = (1.f / m_delta_t)*dens_derive[particle_ind]*alpha[particle_ind];
+
+			pressure_acc_x = pressure_acc_y = pressure_acc_z = 0.0f;
 			
-			div_i = k_v_i[particle_ind] / dens_i;
-			sum_x = sum_y = sum_z = 0.0f;
 			for (auto j = 0; j < m_neighbor_data[particle_ind].n; ++j)
 			{
 				int linear_ind = j + D_MAX_NR_OF_NEIGHBORS*particle_ind;
 				neighbor_ind = m_neighbor_data[particle_ind].neighbor[j];
 
 				assert(m_particles.dens[neighbor_ind] != 0.0f, "n dens");
 
-				dx = m_particles.pos.x[neighbor_ind] - m_particles.pos.x[particle_ind];
-				dy = m_particles.pos.y[neighbor_ind] - m_particles.pos.y[particle_ind];
-				dz = m_particles.pos.z[neighbor_ind] - m_particles.pos.z[particle_ind];
+				k_v_j = (1.f / m_delta_t)*dens_derive[neighbor_ind] * alpha[neighbor_ind];
```

Comment:
1/m_delta_t kan beräknas en gång innan for-looparna för att slippa göras flera ggr. även på rad 340